### PR TITLE
Add worker groups with different settings

### PR DESCRIPTION
**Pull request by severinraez**
A follow up to https://github.com/salsify/delayed_job_worker_pool/pull/3

I've changed the terminology to `worker_groups` so we have clear semantics. The worker pool are all the workers, a worker group is a subset of workers sharing the same options.

### CODE REVIEW

File: lib/delayed_job_worker_pool/dsl.rb [hunk 1]
```diff
@@ -1,29 +1,39 @@
...
-        @options[option_name] = option_value unless option_value.nil?
-      end
+    def preload_app(preload = true)
+      @options[:preload_app] = preload
     end
 
-    def preload_app(preload_app = true)
-      @options[:preload_app] = preload_app
+    def worker_group(&block)
+      pool_options = WorkerGroupOptions.new
```
Comment (jturkel) [inline lib/delayed_job_worker_pool/dsl.rb]: Should this be `group_options` to follow consistent terminology?

File: lib/delayed_job_worker_pool/dsl.rb [hunk 2]
```diff
@@ -1,29 +1,39 @@
...
+      yield(pool_options)
+      @options[:worker_groups] << pool_options
+    end
+
+    def assert_pools_defined!
```
Comment (jturkel) [inline lib/delayed_job_worker_pool/dsl.rb]: Should this be `assert_groups_defined!` to follow consistent terminology?

File: lib/delayed_job_worker_pool/worker_group_options.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+      queues
+      min_priority
+      max_priority
+      sleep_delay
+      read_ahead
+    ].freeze
+    GROUP_SETTINGS = %i[workers].freeze
+    ALL_SETTINGS = (DJ_SETTINGS + GROUP_SETTINGS).freeze
+
+    ALL_SETTINGS.each { |setting| attr_accessor setting }
```
Comment (jturkel) [inline lib/delayed_job_worker_pool/worker_group_options.rb]: Minor but we can shorten this up a bit.

```suggestion
    attr_accessor *ALL_SETTINGS
```

File: lib/delayed_job_worker_pool/worker_pool.rb [hunk 4]
```diff
@@ -3,11 +3,14 @@
...
 module DelayedJobWorkerPool
   class WorkerPool
+
     SIGNALS = ['TERM', 'INT'].map(&:freeze).freeze
+    DEFAULT_WORKER_COUNT = 1
 
     def initialize(options = {})
       @options = options
       @worker_pids = []
+      @worker_pid_options = {}
```
Comment (jturkel) [inline lib/delayed_job_worker_pool/worker_pool.rb]: Can we combine the `@worker_pids` and `@worker_pid_options` data structures to make it easier to keep the information stored in them consistent?

File: lib/delayed_job_worker_pool/worker_pool.rb [hunk 5]
```diff
@@ -114,14 +120,23 @@ def invoke_callback(callback_name, *args)
...
+        worker_options = group.dj_worker_options
+        workers.times { fork_worker(worker_options) }
+      end
+    end
+
+    def fork_worker(options)
+      worker_pid = Kernel.fork { run_worker(options) }
       worker_pids << worker_pid
+      worker_pid_options[worker_pid] = options
       log("Started worker #{worker_pid}")
```
Comment (jturkel) [inline lib/delayed_job_worker_pool/worker_pool.rb]: Would it make sense to give worker groups names so we can include that information in log statements like this to provide visibility into which groups are churning workers?

**Review by jturkel — CHANGES_REQUESTED** [review]
Overall looks good. Can we add some test coverage to `spec/delayed_job_worker_pool_spec.rb` to test the changes to `lib/delayed_job_worker_pool/worker_pool.rb`? I'm also curious to hear your thoughts on providing worker groups with names for better visibility into their state.

**Comment by severinraez** [comment]
Thanks for your review, @jturkel
I agree with all the points you raised. Worker group names will help a lot with knowing what's going on. In that spirit, I plan to also set the `name_prefix` (https://github.com/collectiveidea/delayed_job/blob/master/lib/delayed/worker.rb#L33) of the DJ workers accordingly.

**Comment by severinraez** [comment]
Hi @jturkel 

I've added two commits, handling everything but the worker group names in the first and adding worker group names in the second. Please have a look. I'd be happy to follow up on additional feedback if you have any and then squash the commits to allow for a clean merge.

File: README.md [hunk 6]
```diff
@@ -41,9 +41,11 @@ delayed_job_worker_pool <config file>
 The config file is a Ruby DSL inspired by the [Puma](https://github.com/puma/puma) configuration DSL. Here's an example:
 
 ```ruby
-workers Integer(ENV['NUM_WORKERS'] || 1)
-queues (ENV['QUEUES'] || ENV['QUEUE'] || '').split(',')
-sleep_delay ENV['WORKER_SLEEP_DELAY']
+worker_group(:my_group_name) do |g|
```
Comment (jturkel) [inline README.md]: Should we make the group name optional with a reasonable default under the assumption that a single worker group will still be the most common use case for this gem?

File: lib/delayed_job_worker_pool/registry.rb [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+      raise GroupAlreadyExists, "Group #{group} already exists" if @groups.key?(name)
+
+      @groups[name] = {
+        options: options,
+        pids: []
+      }
+    end
+
+    def add_worker(group_name, pid)
+      @groups[group_name][:pids] << pid
```
Comment (jturkel) [inline lib/delayed_job_worker_pool/registry.rb]: Can we add error handling around a group not being found? Same applies to the `options` method.

File: lib/delayed_job_worker_pool/registry.rb [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    def remove_worker(pid)
+      @groups[group(pid)][:pids].delete(pid)
+    end
+
+    def options(group_name)
+      @groups[group_name][:options]
+    end
+
+    def worker_pids
+      @groups.values.map{ |v| v[:pids] }.flatten
```
Comment (jturkel) [inline lib/delayed_job_worker_pool/registry.rb]:
```suggestion
      @groups.values.flat_map{ |v| v[:pids] }
```

File: lib/delayed_job_worker_pool/registry.rb [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  # Keeps track of worker groups and their workers.
+  class Registry
+    class GroupAlreadyExists < StandardError; end
+    class GroupNotFound < StandardError; end
+
+    def initialize
+      @groups = {}
+    end
+
+    def include?(pid)
```
Comment (jturkel) [inline lib/delayed_job_worker_pool/registry.rb]: Based on the name it wasn't clear to me if this was checking groups or workers. Should we call this `include_worker?` to make the intent more obvious?

File: lib/delayed_job_worker_pool/registry.rb [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+      worker_pids.include?(pid)
+    end
+
+    def empty?
```
Comment (jturkel) [inline lib/delayed_job_worker_pool/registry.rb]: Similar comment as `include?` i.e. it wasn't clear to me if this was checking groups or workers. Should we call this `has_workers?` to make the intent more obvious?

File: spec/delayed_job_worker_pool/registry_spec.rb [hunk 11]
```diff
@@ -0,0 +1,61 @@
+require 'spec_helper'
+
+describe DelayedJobWorkerPool::Registry do
```
Comment (jturkel) [inline spec/delayed_job_worker_pool/registry_spec.rb:3]: Can we add test coverage for edge cases in this class e.g. duplicate group names, looking up workers that don't exists, etc.?

File: spec/delayed_job_worker_pool_spec.rb [hunk 12]
```diff
@@ -75,6 +81,33 @@
...
+      )
+    end
+
+    let(:config_file) { config_file_path('multiple_groups.rb') }
+
+    context 'first group gets its options' do
+      let(:first_group_queues) { 'active' }
+      let(:second_group_queues) { 'other' }
+
+      it_behaves_like 'runs jobs on active queues'
```
Comment (jturkel) [inline spec/delayed_job_worker_pool_spec.rb:100]: Do these tests need to make sure that each group is picking up jobs on the appropriate queues? It looks like it won't detect the scenario where worker groups are listening on all queues.

**Review by jturkel — CHANGES_REQUESTED** [review]
Looking good! I like the extraction of the registry. Mostly minor feedback but the biggest question I have is whether we should have a default worker group name to remove some friction for the common scenario of a single worker pool.

**Comment by severinraez** [comment]
Thanks :) 

I've implemented your suggested changes. I've thought about a default worker group name before and went for explicit namings just to have a decision. You bringing this up too leads me to believe a default might be simpler.

I did not (yet?) add a test to `spec/delayed_job_worker_pool_spec.rb`. I had another look, my thinking here is as follows:

- The existing test makes sure no jobs queued on queue `inactive` are run - thus we can be sure no worker is working all queues. 
- Left unchecked is the following contrived edge case: If we'd somehow build a union of the specified per-worker group queues and passed that to each worker group, the test would be green. In `dsl_spec.rb` we already examine which worker groups get which options.

Thinking further, there's a whole class of checks we don't do around making sure the specified options are really passed through to `Delayed::Worker` (`sleep_delay` and so on). Adding those would bloat `delayed_job_worker_pool_spec.rb`.

I think the proper way to do this would be to separate worker instantiation from forking and looking after children by refactoring `WorkerPool`, then having the worker instantiation part thoroughly tested (maybe by mocking `Delayed::Worker`).

Thus, I suggest we let the test be as is and merge this PR (I can squash the commits first). I can open an issue summarizing the refactoring and opportunities for more thorough testing if you're interested. I'm afraid I'll be too short on time to implement this in the coming months. That being said, if you feel more confident with the "worker working all specified queues" also being tested, I can come up with a solution and add a commit. Might get a little tricky to tell from the outside, we'd need to find out which PID is working which queues from the logs (which would have to be split so the forked processes don't write to the same file and garble the output every now and then, leading to flaky tests).

**Review by jturkel — APPROVED** [review]
I think the test coverage is good enough for now. Thanks for all your work on this. I'll get it merged and released after you rebase/squash.

**Comment by severinraez** [comment]
Thank you too for your guidance, that improved the quality a lot.

I've squashed and rebased :)

**Review by jturkel — APPROVED** [review]
Thanks for the contribution @severinraez! I'll release an updated version of the gem today.
